### ISSUE TITLE
Investigate the use of threadpool loading data

### ISSUE THREAD

**Issue by oysteoh** [equinor/ert#2650]
Issue:

Using threadpool to parallelize loading of realizations with Summary data
For record requests for Summary data, approx 25-40% of time is spent in the function EnsemblePlotData:
https://github.com/equinor/ert/blob/main/res/enkf/export/summary_collector.py#L61

which in turn calls enkf_plot_data_load in libres
https://github.com/equinor/ert/blob/main/libres/lib/enkf/enkf_plot_data.cpp#L108

When loading the data a thread pool is used to load realizations in parallel, but if this code is removed and replaced by serial loading of realizations then the total time is greatly reduced by 90%. This then leads to a total reduction in record request time by between 25-50%

**Comment by BjarneHerland** [comment]
Using drogon-data provided by @TerryHannant  and a C++ driver I can confirm the essence of this issue (which for reference originates from https://github.com/equinor/webviz-ert/issues/197).

The point of using data from drogon is to have enough summary-data to measure load-times properly - I'm not aware of a dataset in our test-toolbox which contains this. The particular dataset contains 142 summary-keys over 553 (active / relevant) realisations, causing the test to load 79520 elements. Not sure why number of elements is `142*560` and not `142*553` - it sums size of all returned `enkf_plot_data_type` - but it indicates the amount of data loaded.

In order to remove (potentially) unnecessary overhead of the legacy thread/threadpool it was replaced with C++ futures from `<async>`. These can launch in three modes: `std::launch::async` schedules jobs immediately (multiple threads), `std::launch::deferred` runs jobs on-demand (essentially loads on the calling thread when looping to collect results), and combination of both modes which leaves scheduling to the system.

The numbers below are from a small number of benchmark-runs for each case, approximated to something close to the average. Note that they are subject to several factors, most notably the speed and load on my laptop when benchmarking, but also the effect of caching ERT-files in memory (see the two footnotes).

Average load-time (ms) per key (approximated over a couple of runs)
```
                                   Native       Python    Overhead
Threadpool, 4 threads (original) :  41(*)       245       204
std::async, launch::deferred     :   6          173       167
std::async, launch::async        : 112          290       178          
std::async, system decides       : 110(**)      290       180
No concurrency                   :   5          171       166

(*) Also saw ~101ms at some point, but could not reproduce this
(**) Also saw ~35ms at some point, but could not reproduce this
```

This indicates that we can safely remove the concurrenty/parallellism in this code, reducing both code-compexity and load-time. There also seems to be some potential for improving the loops in Python but that is another topic.

**Comment by oysteoh** [comment]
Great work @BjarneHerland ! Have you looked into how much work it would be to change the implementation removing the concurrency/parallellism? Is it somewhat limited and easy to change and by that also easy to verify the behaviour in code is the same before / after the change? And also - just out of curiosity - this benchmarking is performed on your local computer? Could it be that it ends up being different on the RGS nodes where we have NFS?

**Comment by BjarneHerland** [comment]
> Have you looked into how much work it would be to change the implementation removing the concurrency/parallellism?

That is trivial and already done in order to provide the numbers above.

> Is it somewhat limited and easy to change and by that also easy to verify the behaviour in code is the same before / after the change?

We could create some snapshot-tests (or convert existing ones), if that is what you mean. The issue may be to clearly see the described behaviour with syntetic data, as these seems to be limited in size. If @mortalisk makes the poly-template he has been working on include summary-data, we should be able to use it.

> And also - just out of curiosity - this benchmarking is performed on your local computer? Could it be that it ends up being different on the RGS nodes where we have NFS?

Definitely. As pointed out by @frode-aarstad, this code is not written by accident. Hence, there must have been some reason for doing it this way. Perhaps @joakim-hove knows this?

**Comment by TerryHannant** [comment]
> Could it be that it ends up being different on the RGS nodes where we have NFS?

I did build and informally test on a RGS node and found that the issue is does exist there as well

**Comment by jondequinor** [comment]
> > Could it be that it ends up being different on the RGS nodes where we have NFS?
> 
> 
> 
> I did build and informally test on a RGS node and found that the issue is does exist there as well 

The work under test was on an NFS disk? I.e. `/private` or `/scratch`?

**Comment by joakim-hove** [comment]
> As pointed out by @frode-aarstad, this code is not written by accident. Hence, there must have been some reason for doing it this way. Perhaps @joakim-hove knows this?

I do not remember this - but the work load here is quite small so I am not surpised the threading does not pay off.

**Comment by BjarneHerland** [comment]
@joakim-hove what would a more realistic workload be in your experience? This is one of the key-issues when deciding how to proceed IMO, in addition to @jondequinor s point about network-disk or not.

**Comment by joakim-hove** [comment]
> what would be a more realistic workload in your experience?

Have to think about that - video next week?

**Comment by BjarneHerland** [comment]
@joakim-hove  any input on this yet? I'm available for a chat at any time, just poke me on slack

**Comment by joakim-hove** [comment]
Sorry about the late reply

> but the work load here is quite small so I am not surpised the threading does not pay off.

I misunderstood at several levels when giving this original comment; I now see that the code in question is *only* about loading summary data from internal storage - that will always be a quite limited workload; I therefor do not see any potential for other models significantly changing the picture. Hence removing the `thread_pool` seems like a nobrainer with both improved performance and reduced complexity - what is not to like?!
